gamestonk_terminal/portfolio/rh_api.py

```diff
@@ -0,0 +1,164 @@
+import argparse
+import datetime
+from datetime import datetime as dt
+import pandas as pd
+from pandas.plotting import register_matplotlib_converters
+import matplotlib.pyplot as plt
+import mplfinance as mpf
+import numpy as np
+
+import robin_stocks.robinhood as r
+from termcolor import colored
+from gamestonk_terminal.helper_funcs import (
+    parse_known_args_and_warn,
+    plot_autoscale,
+)
+from gamestonk_terminal import feature_flags as gtff
+from gamestonk_terminal.config_terminal import RH_USERNAME as user, RH_PASSWORD as pw
+
+register_matplotlib_converters()
+
+dt_format = "%Y-%m-%dT%H:%M:%SZ"
+valid_span = ["day", "week", "month", "3month", "year", "5year", "all"]
+valid_interval = ["5minute", "10minute", "hour", "day", "week"]
+
+span_title_dict = {
+    "day": "Day",
+    "week": "Week",
+    "month": "Month",
+    "3month": "3 Months",
+    "year": "Year",
+    "5year": "5 Years",
+    "all": "All Time",
+}
+
+
+def login():
+    r.login(user, pw)
+
+
+def logoff():
+    r.logout()
+
+
+def show_holdings():
+
+    holds = r.account.build_holdings()
+    stocks = []
+    equity = []
+    for stock, data in holds.items():
+        stocks.append(stock)
+        equity.append(round(float(data["equity"]), 2))
+
+    print("")
+    print("Stonk\t last price \t prev close \t equity \t % Change")
+    print("")
+    for stonk, eq in zip(stocks, equity):
+        stonk_data = r.stocks.get_quotes(stonk)[0]
+        prev_close = round(float(stonk_data["adjusted_previous_close"]), 2)
+        last_price = round(float(stonk_data["last_trade_price"]), 2)
+        pct_change = round((last_price - prev_close) / prev_close, 3)
+        to_print = f"{stonk}\t {last_price}\t\t {prev_close}\t\t {eq}\t\t {pct_change}"
+
+        if last_price >= prev_close:
+            print(colored(to_print, "green"))
+        else:
+            print(colored(to_print, "red"))
+
+    print("")
+
+
+def plot_historical(l_args):
+    parser = argparse.ArgumentParser(
+        add_help=False,
+        prog="Port",
+        description="""
+                            Historical Portfolio Info
+                        """,
+    )
+
+    parser.add_argument(
+        "-s",
+        "--span",
+        dest="span",
+        type=str,
+        default="day",
+        help="Span of historical data",
+    )
+
+    parser.add_argument(
+        "-i",
+        "--interval",
+        dest="interval",
+        default="10minute",
+        type=str,
+        help="Interval to look at portfolio",
+    )
+
+    try:
+        ns_parser = parse_known_args_and_warn(parser, l_args)
+
+        if not ns_parser:
+            return
+
+        if ns_parser.interval not in valid_interval:
```

Comment:
Instead of having it here, you can add this check to the `type`

E.g.
```
parser.add_argument(
        "-s",
        "--start",
        type=valid_date,
        dest="s_start_date",
        help="The starting date (format YYYY-MM-DD) of the stock",
    )
```
where:
```
def valid_date(s: str) -> datetime:
    try:
        return datetime.strptime(s, "%Y-%m-%d")
    except ValueError as value_error:
        raise argparse.ArgumentTypeError("Not a valid date: {s}") from value_error
``´

Reply:
So the example you have there would ensure that -s (stuff) results in stuff having the "%Y-%m-%d" format.

In my case, (I think you are mentioning line 104) I want to check that -i is in a list.  I tried this:

```
def input_in_list(input_val, check_list):
    if input_val in check_list:
        return input_val
    else:
        raise argparse.ArgumentTypeError("Input not in list of valid options")

```

But using type = input_in_list raises nm error (which doesn't surprise me as I am not passing in a list)

Reply:
The issue is that when you use `type=input_in_list` you only use as an argument the input provided by the user, and in that function you have an additional argument called check_list. 

The way to go forward is what I did for the sources of information when loading a stock, i.e. hardcoding the list inside that function. There may be ways to have this type= to accept another argument, but since I knew how to do it with the hardcoded list inside, I went that way.
```
def check_sources(source: str) -> str:
    available_historical_price_sources = ["yf", "av"]
    if source in available_historical_price_sources:
        return source
    raise argparse.ArgumentTypeError(
        "This source for historical data is not available."
    )
```

Reply:
So just curiously, what is the advantage of having that type (if the list is hard coded) vs just checking with the `if _ in _` ?

Reply:
Modularity I would say.

That way all the checks are done at the time of parsing, rather than accepting the arguments, and raising the exception at a later stage.

This will make it easier to decouple functionalities at a later stage, which is something we are keen on looking forward